lib/3scale_toolbox/commands/copy_command.rb

```diff
@@ -0,0 +1,228 @@
+          metric = metrics.find{|metric| metric.fetch('system_name') == copy.fetch('system_name') }
+          metric ||= {}
+
+          [metric['id'], copy['id']]
+        end.to_h
+
+        puts "destroying all mapping rules of the copy which have been created by default"
+        client.list_mapping_rules(service_copy_id).each do |mapping_rule|
+          client.delete_mapping_rule(service_copy_id, mapping_rule['id'])
+        end
+
+        mapping_rules = source_client.list_mapping_rules(service_id)
+        mapping_rules_copy = client.list_mapping_rules(service_copy_id)
+
+        puts "the original service has #{mapping_rules.size} mapping rules"
+        puts "the copy has #{mapping_rules_copy.size} mapping rules"
+
+        unique_mapping_rules_copy = mapping_rules_copy.dup
+
+        missing_mapping_rules = mapping_rules.reject do |mapping_rule|
+          matching_metric = unique_mapping_rules_copy.find do |copy|
+            compare_hashes(mapping_rule, copy, %w(pattern http_method delta)) &&
+                metrics_mapping.fetch(mapping_rule.fetch('metric_id')) == copy.fetch('metric_id')
+          end
+
+          unique_mapping_rules_copy.delete(matching_metric)
+        end
+
+        puts "missing #{missing_mapping_rules.size} mapping rules"
+
+        missing_mapping_rules.each do |mapping_rule|
+          mapping_rule.delete('links')
+          mapping_rule['metric_id'] = metrics_mapping.fetch(mapping_rule.delete('metric_id'))
+          client.create_mapping_rule(service_copy_id, mapping_rule)
+        end
+        puts "created #{missing_mapping_rules.size} mapping rules"
+
+        puts "extra #{unique_mapping_rules_copy.size} mapping rules"
+        puts unique_mapping_rules_copy.each{|rule| rule.delete('links') }
+
+        application_plan_mapping.each do |original_id, copy_id|
+          limits = source_client.list_application_plan_limits(original_id)
+          limits_copy = client.list_application_plan_limits(copy_id)
+
+          missing_limits = limits.reject { |limit| limits_copy.find{|limit_copy| limit.fetch('period') == limit_copy.fetch('period') } }
+
+          missing_limits.each do |limit|
+            limit.delete('links')
+            client.create_application_plan_limit(copy_id, metrics_mapping.fetch(limit.fetch('metric_id')), limit)
+          end
+          puts "copied application plan #{copy_id} is missing #{missing_limits.size} from the original plan #{original_id}"
+        end
+      end
+    end
+
+    module CopyCommand
+      def self.copy_command
+        Cri::Command.define do
+          name        'copy'
+          usage       'copy <command> [options]'
+          summary     '3scale CLI copy'
+          description '3scale CLI copy tools to manage your API from the terminal.'
+
+          flag :h, :help, 'show help for this command' do |_, cmd|
+            puts cmd.help
+            exit 0
+          end
+        end
+      end
+    end
+
+    def self.command_copy_definition
```

Comment:
This pattern is repeated in every command.

It signals few things to me. Looks like this file has more than one responsibility and should be split. And that we should have an example how this split looks like (copy service, copy account, copy provider, each in own file). Finally that there is no way to extend existing commands, just add them to the root.

Reply:
Example how splitting would look like:

```ruby
$ cat lib/3scale_toolbox/commands/copy_command.rb
require 'my_namespace/copy'

module ThreeScaleToolbox
  module Commands
    def self.command_copy_definition
      MYNAMESPACE.copy_command
    end
  end
end

$ cat lib/my_namespace/copy.rb
require 'my_namespace/copy_service'

module MYNAMESPACE
  def self.root_command
     Cri::Command.define do                                                       
       name        'copy'                                                         
       usage       'copy <command> [options]'                                     
       summary     '3scale CLI copy'                                              
       description '3scale CLI copy tools to manage your API from the terminal.'                                                      
     end                                                                          
  end

  def self.copy_command
    root_command.tap { |command| command.add_command copy_service_command  }
  end
end      

$ cat lib/my_namespace/copy_service.rb
module MYNAMESPACE
  def self.copy_service_command
     Cri::Command.define do                                                       
       name        'service'                                                         
       usage       'service [options] <service_id>'                                     
       summary     '3scale CLI copy service'                                              
       description '3scale CLI copy service tools to manage your API from the terminal.'        
       run do |opts, args, cmd|                 
         CopyServiceCommand.run opts, args    
       end 
     end                                                                          
  end
end      
```

Hopefully it is clearer now. I think spliting responsabilities is not a limitation, can be done easilly. I did this way to keep everything in a single file like it was before. Focus on plugin system.

Reply:
I'm sorry but I don't really like that. Imo commands should be classes/modules and not methods. Reopening a module to add a method to some global bucket when you can just define class and give it a name feels wrong to me. IMHO Class/Module should be defined in one file (except for case when modules are used as plain namespaces with no methods) and this violates that.

For the sake of this example I think the "copy" command and its subcommands will be shipped with the toolbox, so they should use the same namespace.

```
lib/3scale_toolbox/commands/copy_command.rb
lib/3scale_toolbox/commands/copy_command/copy_service.rb
lib/3scale_toolbox/commands/copy_command/copy_account.rb
lib/3scale_toolbox/commands/copy_command/copy_provider.rb
```

Another beef that I start to see is the naming juggling. Changing the name of the method that defines the command is super confusing and hard to follow. I'd definitely like more just constants and some well defined API. I understand that you need to chain the definitions, but I think that could be improved to be more idiomatic. Just provide some API to do that.

Here is a few alternatives.

```ruby
module ThreeScaleToolbox::Commands::CopyCommand
  def self.command
     Cri::Command.define do  
       ...
     end
  end

  def self.subcommands
    [ CopyService ] 
  end
end

module ThreeScaleToolbox::Commands::CopyCommand::CopyService
  def self.command
    Cri::Command.define do 
      ...
    end
  end
end
```


```ruby
module ThreeScaleToolbox::Commands::CopyCommand
  SUBCOMMANDS = []

  def self.extended(subcommand)
    SUBCOMMANDS << subcommand
  end

  def self.command
     Cri::Command.define do  
       ...
     end
  end

  def self.subcommands
    SUBCOMMANDS
  end
end

module ThreeScaleToolbox::Commands::CopyCommand::CopyService
  extend ThreeScaleToolbox::Commands::CopyCommand

  def self.command
    Cri::Command.define do 
      ...
    end
  end
end
```


```ruby
module ThreeScaleToolbox::Commands::CopyCommand
  SUBCOMMANDS = []

  def self.extended(subcommand)
    SUBCOMMANDS << subcommand
  end

  def self.command
     Cri::Command.define do  
       ...
     end.tap do |command|
       SUBCOMMANDS.each(&command.method(:add_command))
     end
  end
end

module ThreeScaleToolbox::Commands::CopyCommand::CopyService
  extend ThreeScaleToolbox::Commands::CopyCommand

  def self.command
    Cri::Command.define do 
      ...
    end
  end
end
```

I'm not saying these are great or better. Just some food for thought.